### PR TITLE
[fix]: [CCM-8934]: Creating multiple BMs if click Save Cost Category multiple times instantly

### PR DESCRIPTION
**Pull request by iamanmolm**
## Describe your changes
On creating a new business mapping/cost category, we query the db to find out if there is any existing business mapping/cost category with the same name. Since right now with current implementation we can have list of such business mapping, we are returning the business mapping as list.

If it exists we throw an InvalidRequestException("Cost category name already exists.").

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39293)

### CODE REVIEW

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 1]
```diff
@@ -103,6 +109,13 @@ public RestResponse<String> update(
       @QueryParam(NGCommonEntityConstants.ACCOUNT_KEY) String accountId, BusinessMapping businessMapping) {
     rbacHelper.checkCostCategoryEditPermission(accountId, null, null);
     BusinessMapping oldCostCategory = businessMappingService.get(businessMapping.getUuid(), accountId);
+    if (!oldCostCategory.getName().equals(businessMapping.getName())) {
+      List<BusinessMapping> existingMapping = businessMappingService.getByName(businessMapping.getName(), accountId);
+      if (existingMapping != null && existingMapping.size() > 0) {
+        log.info("Ce-dev: Old BM Name:" + oldCostCategory.getName() + "and new BM name:" + businessMapping.getName());
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:115]: Here, we can change the log type to `warn`. And, please remove `Ce-dev: `  from the log line.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 2]
```diff
@@ -103,6 +109,13 @@ public RestResponse<String> update(
...
+        throw new InvalidRequestException("Cost category name already exists.");
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:116]: How are we handling this exception at the UI side?

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java [hunk 3]
```diff
@@ -103,6 +109,13 @@ public RestResponse<String> update(
       @QueryParam(NGCommonEntityConstants.ACCOUNT_KEY) String accountId, BusinessMapping businessMapping) {
     rbacHelper.checkCostCategoryEditPermission(accountId, null, null);
     BusinessMapping oldCostCategory = businessMappingService.get(businessMapping.getUuid(), accountId);
+    if (!oldCostCategory.getName().equals(businessMapping.getName())) {
+      List<BusinessMapping> existingMapping = businessMappingService.getByName(businessMapping.getName(), accountId);
+      if (existingMapping != null && existingMapping.size() > 0) {
```
Comment (sdsingh07) [inline ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:114]: We can also use `Lists.isNullOrEmpty` function.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java (hunk 1 above)
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:115]: Yes, I will be removing it. I was just testing it out.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java (hunk 2 above)
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:116]: Right now, it comes as "Cost category name already exists.". UI Changes side changes with this & maybe disabling the save cost category button are still to be done.

File: ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java (hunk 3 above)
Comment (iamanmolm) [reply ce-nextgen/service/src/main/java/io/harness/ccm/remote/resources/BusinessMappingResource.java:114]: Sure, using it.

**Review by sdsingh07 — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39293#-:-NFSI1MI0fDKP7JSR_5Y:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**